sproxy.go

```diff
@@ -0,0 +1,189 @@
+	ResponseSize int64     `json:"res_size,omitempty"`
+	StartTime    time.Time `json:"start,omitempty"`
+	EndTime      time.Time `json:"end,omitempty"`
+	ID           string    `json:"id,omitempty"`
+	Err          error     `json:"err,omitempty"`
+}
+
+type MetricsReporter struct {
+	currentEvents []*Event
+	period        time.Duration
+
+	cancelChan <-chan bool
+	cancelFn   func() error
+
+	// mu protects statsRecvFn
+	mu          sync.RWMutex
+	statsRecvFn func(*InflightStats) error
+}
+
+const defaultPeriod = 1 * time.Minute
+
+func NewMetricsReporter(period time.Duration) (*MetricsReporter, error) {
+	if period <= 0 {
+		period = defaultPeriod
+	}
+	cancelChan, cancelFn := makeCanceler()
+	mr := &MetricsReporter{
+		period:        period,
+		currentEvents: make([]*Event, 0, 512), // Arbitrary value
+		cancelChan:    cancelChan,
+		cancelFn:      cancelFn,
+	}
+	return mr, nil
+}
+
+func (mr *MetricsReporter) Close() error {
+	if mr.cancelFn == nil {
+		return nil
+	}
+	return mr.cancelFn()
+}
+
+func (mr *MetricsReporter) AddEvent(evt *Event) {
+	mr.mu.Lock()
+	mr.currentEvents = append(mr.currentEvents, evt)
+	mr.mu.Unlock()
+}
+
+var errAlreadyClosed = errors.New("already closed")
+
+func makeCanceler() (<-chan bool, func() error) {
+	cancelChan := make(chan bool, 1)
+	var cancelOnce sync.Once
+	cancelFn := func() error {
+		var err error = errAlreadyClosed
+		cancelOnce.Do(func() {
+			close(cancelChan)
+			err = nil
+		})
+		return err
+	}
+
+	return cancelChan, cancelFn
+}
+
+// Do is a synchronous method that reports events
+// at the end of every specified sampling period
+// sending them to the StackDriver Monitoring API.
+func (mr *MetricsReporter) Do() {
+	startTime := time.Now()
+	i := uint64(0)
+	for {
+		select {
+		case <-mr.cancelChan:
+			return
+		case <-time.After(mr.period):
+			// Otherwise the duration has lapsed
+		}
+
+		i += 1
+		drainedEvents := mr.drainEvents()
+
+		go mr.reportEvents(drainedEvents, startTime, time.Now(), i)
+	}
+}
+
+// drainEvents is a concurrent-safe method that
+// creates a frozen copy of mr.currentEvents and
+// then drains mr.currentEvents resetting it an empty slice.
+func (mr *MetricsReporter) drainEvents() []*Event {
+	mr.mu.Lock()
+	drained := make([]*Event, len(mr.currentEvents))
+	copy(drained, mr.currentEvents)
+	mr.currentEvents = mr.currentEvents[:0]
+	mr.mu.Unlock()
+	return drained
+}
+
+func (mr *MetricsReporter) reportEvents(events []*Event, startTime, endTime time.Time, nIter uint64) {
+	nResponses := 0
+	nRequests := 0
+	totalResponseDuration := time.Duration(0)
+	responsesCounter := make(map[int]int)
+	idList := make([]string, 0, len(events))
+	for _, evt := range events {
+		idList = append(idList, evt.ID)
+		switch evt.Request {
+		case true:
+			nRequests += 1
+		default:
+			statusIndex := evt.StatusCode / 100
+			responsesCounter[statusIndex] += 1
+			nResponses += 1
+			totalResponseDuration += evt.EndTime.Sub(evt.StartTime)
+		}
+	}
+
+	// Then for the final report
```

Comment:
then what?

Reply:
haha my bad, vestige of old code, removing it.